### PR TITLE
Fix multi-asic behaviour for dropstat

### PR DESCRIPTION
**Pull request by bktsim-arista**
#### What I did
Previously, dropstat was not behaving correctly on multi-asic devices as the correct namespaces were not being traversed and namespace option `-n` was not available. This change fixes the multi-asic behaviour of this command and adds the `-n` option, with relevant unit tests to ensure correct functionality.

This is a part of the set of changes being pushed for https://github.com/sonic-net/sonic-buildimage/issues/15148
#### How I did it
Added `-n` argument for multi-asic devices and relevant unit tests.

#### How to verify it
Use `dropstat -c show` with `-n`.

### CODE REVIEW

File: scripts/dropstat [hunk 1]
```diff
@@ -403,15 +408,25 @@ Examples:
...
     counter_type = args.type
+    namespace = args.namespace
 
-    dcstat = DropStat()
+    load_db_config()
+    namespaces = multi_asic.get_namespace_list()
+    if namespace and namespace not in namespaces:
+        raise Exception("Input arguments error. Namespaces must be one of", *namespaces)
+
+    if multi_asic.is_multi_asic() and not namespace:
```
Comment (wenyiz2021) [inline scripts/dropstat:426]: this can be replaced with a more consistent way at line 411:
@click.option('-n', '--namespace', help='Namespace name',
             required=True if multi_asic.is_multi_asic() else False, type=click.Choice(multi_asic.get_namespace_list()))

File: tests/multi_asic_dropstat_test.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+class TestMultiAsicDropstat(object):
+   @classmethod
+   def setup_class(cls):
+      if os.path.exists(dropstat_path):
+         shutil.rmtree(dropstat_path)
+      os.environ["PATH"] += os.pathsep + scripts_path
+      os.environ["UTILITIES_UNIT_TESTING"] = "1"
+      os.environ["UTILITIES_UNIT_TESTING_TOPOLOGY"] = "multi_asic"
+      print("SETUP")
+
```
Comment (wenyiz2021) [inline tests/multi_asic_dropstat_test.py:36]: can you add edge cases like:
1. when no namespace is given for multi-asic
2. when namespace is given on single-asic

**Comment by vmittal-msft** [comment]
@bktsim-arista Few things -
1. Can you please share how "Show dropcounter capabilites/configuration/counts" command is going to change ? 
2. please add test for all options under dropcounter.
3. If no namespace specified, Will it show for all namespaces? if we specify, will it show for only that namespace ?
4. Is this verified to be working fine on pizza box as well as chassis?

**Review by vmittal-msft — COMMENTED** [review]
Please check comment

File: scripts/dropstat (hunk 1 above)
Comment (arista-nwolfe) [reply scripts/dropstat]: @wenyiz2021 are you sure we should be using the `@click` decorators within the `scripts/` folder?
This `dropstat` file doesn't hold any CLI commands just the implementation code for querying the redis DBs

**Comment by arista-nwolfe** [comment]
> @bktsim-arista Few things -
> 
> 1. Can you please share how "Show dropcounter capabilites/configuration/counts" command is going to change ?
> 2. please add test for all options under dropcounter.
> 3. If no namespace specified, Will it show for all namespaces? if we specify, will it show for only that namespace ?
> 4. Is this verified to be working fine on pizza box as well as chassis?

For questions 1 & 2: This PR only adds multi-asic support to `show dropcounter counts` so we shouldn't include test coverage for `capabilities/configuration` as it's not supported yet on multi-asic.

File: scripts/dropstat (hunk 1 above)
Comment (kenneth-arista) [reply scripts/dropstat]: See latest change. I took your suggestion to use the click library. Thanks!

File: tests/multi_asic_dropstat_test.py (hunk 2 above)
Comment (kenneth-arista) [reply tests/multi_asic_dropstat_test.py]: See latest change. Above two edges have been covered in new UTs.

**Comment by kenneth-arista** [comment]
@arlakshm please review

File: scripts/dropstat [hunk 3]
```diff
@@ -457,39 +482,22 @@ class DropStat(object):
...
-
-    args = parser.parse_args()
-
-    command = args.command
-
-    group = args.group
-    counter_type = args.type
-
-    dcstat = DropStat()
+@click.command(help='Display drop counters')
```
Comment (arlakshm) [inline scripts/dropstat:485]: why do we need to add this change here?
`dropstat` is standalone python script. the click options need to present in `show/dropcounter.py`

File: show/dropcounters.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 import click
 import utilities_common.cli as clicommon
+import utilities_common.multi_asic as multi_asic_util
```
Comment (arlakshm) [inline show/dropcounters.py:3]: Can you add some test to cover the single asic as well?

File: scripts/dropstat (hunk 3 above)
Comment (kenneth-arista) [reply scripts/dropstat:485]: The original script also supported argument parsing. Using the `click` simplified that logic significantly. It was suggested earlier by Wenyi and I agree this is an improvement over what was originally implemented. Furthermore, the UTs directly call this script.

**Comment by kenneth-arista** [comment]
> * If no namespace specified, Will it show for all namespaces? if we specify, will it show for only that namespace ?

The latest changes supports this.

**Review by wenyiz2021 — APPROVED** [review]
LGTM

**Comment by wenyiz2021** [comment]
@arlakshm @vmittal-msft please check if your review comments are resolved

**Comment by kenneth-arista** [comment]
> 4. Is this verified to be working fine on pizza box as well as chassis?
Yes, I verified that the changes work on a pizza box with a single ASIC. 

For example, 

```
admin@pkz303:/$ show dropcounters counts  --help
Usage: show dropcounters counts [OPTIONS]

  Show drop counts

Options:
  -g, --group TEXT
  -t, --counter_type TEXT
  --verbose                Enable verbose output
  -n, --namespace []       Namespace name or all
  -h, -?, --help           Show this message and exit.

admin@pkz303:/$ show dropcounters counts
     IFACE    STATE    RX_ERR    RX_DROPS    TX_ERR    TX_DROPS    DEBUG1    DEBUG2    DEBUG3
----------  -------  --------  ----------  --------  ----------  --------  --------  --------
 Ethernet0        U         0           0         0           0         0         0         0
 Ethernet1        X         0           0         0           0         0         0         0
 Ethernet2        U         0           0         0           0         0         0         0
 Ethernet3        U         0           0         0           0         0         0         0
 Ethernet4        X         0           0         0           0         0         0         0
 Ethernet5        X         0           0         0           0         0         0         0
 Ethernet6        X         0           0         0           0         0         0         0
 Ethernet7        U         0           0         0           0         0         0         0
 Ethernet8        U         0           0         0           0         0         0         0
 Ethernet9        U         0           0         0           0         0         0         0
Ethernet10        U         0           0         0           0         0         0         0
Ethernet11        X         0           0         0           0         0         0         0
Ethernet12        U         0           0         0           0         0         0         0
Ethernet13        U         0           0         0           0         0         0         0
Ethernet14        U         0           0         0           0         0         0         0
Ethernet15        U         0           0         0           0         0         0         0
Ethernet16        X         0           0         0           0         0         0         0
Ethernet17        X         0           0         0           0         0         0         0
Ethernet18        U         0           0         0           0         0         0         0
Ethernet19        X         0           0         0           0         0         0         0
Ethernet20        U         0           0         0           0         0         0         0
Ethernet21        X         0           0         0           0         0         0         0
Ethernet22        U         0           0         0           0         0         0         0
Ethernet23        U         0           0         0           0         0         0         0
Ethernet24        U         0           0         0           0         0         0         0
Ethernet25        U         0           0         0           0         0         0         0
Ethernet26        U         0           0         0           0         0         0         0
Ethernet27        U         0           0         0           0         0         0         0
Ethernet28        U         0           0         0           0         0         0         0
Ethernet29        X         0           0         0           0         0         0         0
Ethernet30        U         0           0         0           0         0         0         0
Ethernet31        X         0           0         0           0         0         0         0
Ethernet32        U         0           0         0           0         0         0         0
Ethernet33        U         0           0         0           0         0         0         0
Ethernet34        X         0           0         0           0         0         0         0
Ethernet35        U         0           0         0           0         0         0         0
Ethernet36        U         0           0         0           0         0         0         0
Ethernet37        X         0           0         0           0         0         0         0
Ethernet38        U         0           0         0           0         0         0         0
Ethernet39        X         0           0         0           0         0         0         0
Ethernet40        X         0           0         0           0         0         0         0
Ethernet41        U         0           0         0           0         0         0         0
Ethernet42        U         0           0         0           0         0         0         0
Ethernet43        X         0           0         0           0         0         0         0
Ethernet44        X         0           0         0           0         0         0         0
Ethernet45        U         0           0         0           0         0         0         0
Ethernet46        X         0           0         0           0         0         0         0
Ethernet47        U         0           0         0           0         0         0         0
Ethernet48        U         0           0         0           0         0         0         0
Ethernet49        U         0           0         0           0         0         0         0
Ethernet50        U         0           0         0           0         0         0         0
Ethernet51        U         0           0         0           0         0         0         0

admin@pkz303:/$ show dropcounters counts -g TEST

     IFACE    STATE    DEBUG3
----------  -------  --------
 Ethernet0        U         0
 Ethernet1        X         0
 Ethernet2        U         0
 Ethernet3        U         0
 Ethernet4        X         0
 Ethernet5        X         0
 Ethernet6        X         0
 Ethernet7        U         0
 Ethernet8        U         0
 Ethernet9        U         0
Ethernet10        U         0
Ethernet11        X         0
Ethernet12        U         0
Ethernet13        U         0
Ethernet14        U         0
Ethernet15        U         0
Ethernet16        X         0
Ethernet17        X         0
Ethernet18        U         0
Ethernet19        X         0
Ethernet20        U         0
Ethernet21        X         0
Ethernet22        U         0
Ethernet23        U         0
Ethernet24        U         0
Ethernet25        U         0
Ethernet26        U         0
Ethernet27        U         0
Ethernet28        U         0
Ethernet29        X         0
Ethernet30        U         0
Ethernet31        X         0
Ethernet32        U         0
Ethernet33        U         0
Ethernet34        X         0
Ethernet35        U         0
Ethernet36        U         0
Ethernet37        X         0
Ethernet38        U         0
Ethernet39        X         0
Ethernet40        X         0
Ethernet41        U         0
Ethernet42        U         0
Ethernet43        X         0
Ethernet44        X         0
Ethernet45        U         0
Ethernet46        X         0
Ethernet47        U         0
Ethernet48        U         0
Ethernet49        U         0
Ethernet50        U         0
Ethernet51        U         0

admin@pkz303:/$ show dropcounters counts -t PORT_INGRESS_DROPS
     IFACE    STATE    RX_ERR    RX_DROPS    DEBUG1    DEBUG2    DEBUG3
----------  -------  --------  ----------  --------  --------  --------
 Ethernet0        U         0           0         0         0         0
 Ethernet1        X         0           0         0         0         0
 Ethernet2        U         0           0         0         0         0
 Ethernet3        U         0           0         0         0         0
 Ethernet4        X         0           0         0         0         0
 Ethernet5        X         0           0         0         0         0
 Ethernet6        X         0           0         0         0         0
 Ethernet7        U         0           0         0         0         0
 Ethernet8        U         0           0         0         0         0
 Ethernet9        U         0           0         0         0         0
Ethernet10        U         0           0         0         0         0
Ethernet11        X         0           0         0         0         0
Ethernet12        U         0           0         0         0         0
Ethernet13        U         0           0         0         0         0
Ethernet14        U         0           0         0         0         0
Ethernet15        U         0           0         0         0         0
Ethernet16        X         0           0         0         0         0
Ethernet17        X         0           0         0         0         0
Ethernet18        U         0           0         0         0         0
Ethernet19        X         0           0         0         0         0
Ethernet20        U         0           0         0         0         0
Ethernet21        X         0           0         0         0         0
Ethernet22        U         0           0         0         0         0
Ethernet23        U         0           0         0         0         0
Ethernet24        U         0           0         0         0         0
Ethernet25        U         0           0         0         0         0
Ethernet26        U         0           0         0         0         0
Ethernet27        U         0           0         0         0         0
Ethernet28        U         0           0         0         0         0
Ethernet29        X         0           0         0         0         0
Ethernet30        U         0           0         0         0         0
Ethernet31        X         0           0         0         0         0
Ethernet32        U         0           0         0         0         0
Ethernet33        U         0           0         0         0         0
Ethernet34        X         0           0         0         0         0
Ethernet35        U         0           0         0         0         0
Ethernet36        U         0           0         0         0         0
Ethernet37        X         0           0         0         0         0
Ethernet38        U         0           0         0         0         0
Ethernet39        X         0           0         0         0         0
Ethernet40        X         0           0         0         0         0
Ethernet41        U         0           0         0         0         0
Ethernet42        U         0           0         0         0         0
Ethernet43        X         0           0         0         0         0
Ethernet44        X         0           0         0         0         0
Ethernet45        U         0           0         0         0         0
Ethernet46        X         0           0         0         0         0
Ethernet47        U         0           0         0         0         0
Ethernet48        U         0           0         0         0         0
Ethernet49        U         0           0         0         0         0
Ethernet50        U         0           0         0         0         0
Ethernet51        U         0           0         0         0         0

admin@pkz303:/$ dropstat --version
dropstat, version 1.0
```

**Comment by wenyiz2021** [comment]
@arlakshm has a concern on using `click` option in `scripts` directory which is outside of `show` directory.
otherwise LGTM.
thanks @kenneth-arista for adding test as well

**Review by wenyiz2021 — APPROVED** [review]

**Comment by kenneth-arista** [comment]
> @arlakshm has a concern on using `click` option in `scripts` directory which is outside of `show` directory. otherwise LGTM. thanks @kenneth-arista for adding test as well

Thanks @wenyiz2021 . The @click library usage in the script significantly improves the readability and maintainability of the standalone script. Thanks Wenyi for suggesting it. I think this is a good change. @arlakshm please let me know how you would like to proceed. I recommend keeping the change 😄

**Comment by kenneth-arista** [comment]
@wenyiz2021 can you help merge this change?
